src/mscorlib/shared/System/SpanHelpers.Char.cs

```diff
@@ -0,0 +1,201 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+#if !netstandard
+using Internal.Runtime.CompilerServices;
+#endif
+
+#if !netstandard11
+using System.Numerics;
+#endif
+
+namespace System
+{
+    internal static partial class SpanHelpers
+    {
+        public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref char second, int secondLength)
+        {
+            Debug.Assert(firstLength >= 0);
+            Debug.Assert(secondLength >= 0);
+
+            int lengthDelta = firstLength - secondLength;
+
+            if (Unsafe.AreSame(ref first, ref second))
+                goto Equal;
+
+            IntPtr minLength = (IntPtr)((firstLength < secondLength) ? firstLength : secondLength);
+            IntPtr i = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+
+            if ((byte*)minLength >= (byte*)(sizeof(UIntPtr) / sizeof(char)))
+            {
+#if !netstandard11
+                if (Vector.IsHardwareAccelerated && (byte*)minLength >= (byte*)Vector<ushort>.Count)
+                {
+                    IntPtr nLength = minLength - Vector<ushort>.Count;
+                    do
+                    {
+                        if (Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) !=
+                            Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
+                        {
+                            break;
+                        }
+                        i += Vector<ushort>.Count;
+                    }
+                    while ((byte*)nLength >= (byte*)i);
+                }
+#endif
+
+                while ((byte*)minLength >= (byte*)(i + sizeof(UIntPtr) / sizeof(char)))
```

Comment:
Could this overflow? `(i + sizeof(UIntPtr) / sizeof(char))`

Reply:
It can technically overflow, but it does matter. The comparison is done as `byte*` that is unsigned. Ideally, we would just use nuint here ... but then we would not be able to compile this code in portable System.Memory without major perf hit.